Add Embedded Swift update blog post (#1236)
apple/swift-org-website: _posts/2025-11-17-embedded-swift-improvements-coming-in-swift-6.3.md

---
layout: new-layouts/post
published: true
date: 2025-11-17 10:00:00
title: "Embedded Swift Improvements Coming in Swift 6.3"
author: [douglas_gregor, rauhul]
category: "Adopters"
---

[Embedded Swift](https://www.swift.org/get-started/embedded/) is a subset of Swift that can be used to develop bare-metal, embedded, and standalone programs for a variety of hardware platforms and environments. Embedded Swift is currently an experimental feature that uses a special compilation mode to produce significantly smaller binaries than regular Swift, using techniques described in the [Embedded Swift vision document](https://github.com/swiftlang/swift-evolution/blob/main/visions/embedded-swift.md).

> Embedded Swift is evolving rapidly. This post describes a number of improvements we’ve made in the last few months, covering everything from improved C interoperability to better debugging and steps toward a complete linkage model for Embedded Swift. These features and bug fixes will be available in the upcoming [Swift 6.3](https://forums.swift.org/t/swift-6-3-release-process/82843) release, but you can try them out today with a [Swift development snapshot](https://www.swift.org/install/).

Reviewer: ```suggestion
Embedded Swift is evolving rapidly. This post describes a number of improvements made in the last few months, covering everything from improved C interoperability to better debugging and steps toward a complete linkage model for Embedded Swift.
These features and bug fixes are included in the upcoming [Swift 6.3](https://forums.swift.org/t/swift-6-3-release-process/82843) release, and you can try them out today with a [Swift development snapshot](https://www.swift.org/install/).
```

Slight rewording suggestion, leaning on James' suggestions. It's still passive voice, but I think inferring the people contributing to Embedded Swift work is pretty reasonable here.

## Libraries and Diagnostics

### Printing floating point numbers

> Previously, the `description` and `debugDescription` properties for floating-point types (`Float`, `Double`, etc.) were not available in the Embedded Swift standard library. With a new [all-Swift implementation](https://github.com/swiftlang/swift/pull/84826), these are now available.

Reviewer: ```suggestion
Previously, the `description` and `debugDescription` properties for floating-point types (`Float`, `Double`, etc.) were not available in the Embedded Swift standard library. With a new [all-Swift implementation](https://github.com/swiftlang/swift/pull/84826), you can use this with Embedded Swift.
```

### Embedded restriction diagnostics

> There is a new opt-in set of warnings in the [EmbeddedRestrictions](https://docs.swift.org/compiler/documentation/diagnostics/embedded-restrictions) diagnostic group that diagnoses language constructs that aren’t available in Embedded Swift, such as uses of untyped throws or calling generic functions on existential values (follow the earlier link for more information). These diagnostics are enabled by default when building Embedded Swift, and can also be enabled in non-Embedded Swift using the compiler flag `-Wwarning EmbeddedRestrictions`.

Reviewer 2: The '(follow the earlier link for more information).' seems a little awkward.

Possibly move the link to the parenthetical and update it to something like:

'(See [documentation](https://docs.swift.org/compiler/documentation/diagnostics/embedded-restrictions) for full details.)'

> There is a new opt-in set of warnings in the [EmbeddedRestrictions](https://docs.swift.org/compiler/documentation/diagnostics/embedded-restrictions) diagnostic group that diagnoses language constructs that aren’t available in Embedded Swift, such as uses of untyped throws or calling generic functions on existential values (follow the earlier link for more information). These diagnostics are enabled by default when building Embedded Swift, and can also be enabled in non-Embedded Swift using the compiler flag `-Wwarning EmbeddedRestrictions`.

Reviewer: ```suggestion
There is a new opt-in set of warnings in the [EmbeddedRestrictions](https://docs.swift.org/compiler/documentation/diagnostics/embedded-restrictions) diagnostic group that diagnoses language constructs that aren’t available in Embedded Swift, such as uses of untyped throws or calling generic functions on existential values.
These diagnostics are enabled by default when building Embedded Swift, and can also be enabled in non-Embedded Swift using the compiler flag `-Wwarning EmbeddedRestrictions`.
Read the full details at [EmbeddedRestrictions](https://docs.swift.org/compiler/documentation/diagnostics/embedded-restrictions).
```

// FIXME: show the package manifest syntax for enabling this.

### Swift MMIO 0.1.x

> The [0.1.x release](https://github.com/apple/swift-mmio/releases/tag/0.1.0) of Swift MMIO brings a bunch of bug fixes and quality-of-life improvements, plus brand new comprehensive [documentation](https://swiftpackageindex.com/apple/swift-mmio/0.1.1/documentation/mmio) on the Swift Package Index.

Reviewer: ```suggestion
The [0.1.x release](https://github.com/apple/swift-mmio/releases/tag/0.1.0) of Swift MMIO brings a bunch of bug fixes and quality-of-life improvements, plus brand new comprehensive [documentation](https://swiftpackageindex.com/apple/swift-mmio/documentation/mmio) on the Swift Package Index.
```

Eliding the version of the docs will make the docs link "last" a bit better as the page evolves. SPI always provides a link to the latest release available when you don't provide an explicit one.

> The biggest addition is code generation support. There's now an [svd2swift tool](https://swiftpackageindex.com/apple/swift-mmio/0.1.1/documentation/svd2swift) and corresponding [SwiftPM plugin](https://swiftpackageindex.com/apple/swift-mmio/0.1.1/documentation/svd2swift/usingsvd2swiftplugin) that generates Swift MMIO interfaces directly from CMSIS System View Description (SVD) files. You can run it manually from the command line, or configure the plugin to handle everything automatically at build time.

Reviewer: ```suggestion
The biggest addition is code generation support. There's now an [svd2swift tool](https://swiftpackageindex.com/apple/swift-mmio/documentation/svd2swift) and corresponding [SwiftPM plugin](https://swiftpackageindex.com/apple/swift-mmio/documentation/svd2swift/usingsvd2swiftplugin) that generates Swift MMIO interfaces directly from CMSIS System View Description (SVD) files. You can run it manually from the command line, or configure the plugin to handle everything automatically at build time.
```

Same trick with eliding the version of the package from the links. Is `CMSIS` a well known acronym in embedded circles? I didn't know what it was - do you think it's worth spelling it out (it's awfully wordy), or providing a good neutral target link to help illuminate curious readers?

Author: I have it spelled out in the MMIO docs and yeah it's extremely long. IMO there are no good intro docs to link to for SVD so I just omitted linking anything as it's all technical.

> Debugging also got a nice upgrade with [SVD2LLDB](https://swiftpackageindex.com/apple/swift-mmio/0.1.1/documentation/svd2lldb). This LLDB plugin lets you work with device registers using their actual names instead of fumbling with raw memory addresses. It even includes visual decoding support to help you make sense of register values. For example, here's what it looks like when you decode a timer control register:

Reviewer: ```suggestion
Debugging also got a nice upgrade with [SVD2LLDB](https://swiftpackageindex.com/apple/swift-mmio/documentation/svd2lldb).
This LLDB plugin lets you work with device registers using their actual names instead of fumbling with the raw memory addresses.
It even includes visual decoding support to help you make sense of register values.
For example, here's what it looks like when you decode a timer control register:
```

The `@c` attribute formalizes the `@_cdecl` attribute, fixing a number of corner cases in the process.

### Improved tolerance of mismatching C signatures

The Swift compiler has historically required precise consistency between the Swift types for C functions imported from headers or declared in Swift with `@c` (formerly `@_cdecl`) or `@_extern(c)`. For example, if a C function were declared without nullability annotations in a header:

```c
void takePointer(const double *);
```

And then separately defined in Swift with, e.g.,

```swift
@c func takePointer(_ values: UnsafePointer<Double>) { ... }
```

The compiler could fail to compile when it noticed the mismatch, often with a hard-to-understand “deserialization” failure. The compiler now keeps the different views on the underlying C declaration separate, diagnosing problems when the underlying C declarations themselves are inconsistent. This allows subtle differences such as nullability or sendability annotations to occur in C signatures without failing to compile.

## Debugging 

### Debugger printing of values

> LLDB’s support for printing values of Swift types in Embedded Swift has been improved. In addition, the `memory read` command now supports an optional Swift type name, so that you can take an arbitrary address and render it as a value of the named Swift type with:

Reviewer: ```suggestion
LLDB’s support for printing the values of Swift types in Embedded Swift has been improved. In addition, the `memory read` command now supports an optional Swift type name, so that you can take an arbitrary address and render it as a value of that named Swift type.
The following example shows interpreting the address as the type `MyMessage`:
```

```swift
(lldb) memory read -t MyMessage 0x000000016fdfe970
(MessageLib.MyMessage) 0x16fdfe970 = {
  id = 15
  timestamp = 793219406
  payload = "hello"
  ...
}
```

### More data types inspectable in core dumps

> LLDB needs access to type layout information so it can display variables. Since Embedded Swift doesn't contain reflection metadata, the Swift compiler emits all the type layout information as DWARF debug info. There have been several improvements to Swift compiler's debug info output recently, such as support for type declarations nested inside an extension. At the same time LLDB added support for nested generic type aliases in Embedded Swift. These two improvements together make it possible to inspect many common standard library data types, such as `Dictionary` and `Array`, in Embedded Swift core dumps. Previously these data types were only accessible via the expression evaluator, which requires a live process. 

Reviewer: ```suggestion
LLDB needs access to type layout information so it can display variables. Since Embedded Swift doesn't contain reflection metadata, the Swift compiler emits all the type layout information as DWARF debug info.
There have been several improvements to the Swift compiler's debug info output, such as support for type declarations nested inside an extension.
At the same time LLDB added support for nested generic type aliases in Embedded Swift.
These two improvements together make it possible to inspect many common standard library data types, such as `Dictionary` and `Array`, in Embedded Swift core dumps.
Previously these data types were only accessible via the expression evaluator, which requires a live process. 
```

LLDB also has native support for the new `InlineArray` data type.

### Debugger armv7m exception frames unwinding

LLDB's support for producing backtraces after taking exceptions in armv7m has been greatly improved. Previously after taking an exception users would be presented with a back trace like the following:

```swift
(lldb) bt
* thread 1
  * frame #0: 0x0020392c NeoPixelRainbow`NeoPixelRainbow.swift_UsageFault_Handler() -> () + 28
    frame #1: 0x00203910 NeoPixelRainbow`UsageFault_Handler + 8
    frame #2: 0xfffffff8
    frame #3: 0x00202a86 NeoPixelRainbow`NeoPixelRainbow_main + 8
    frame #4: 0x00200256 NeoPixelRainbow`cinit_ctor_loop_end at startup.S:98
    frame #5: 0x00200210 NeoPixelRainbow`Reset_Handler at startup.S:33
```

However this back trace would often be missing one or more frames before the start of the exception frame (`UsageFault_Handler` above).

> Now, LLDB is able to walk back through exception frames into the standard program frames and produce a backtrace containing the missing frame(s).

Reviewer: ```suggestion
Now, LLDB is able to walk back through exception frames into the standard program frames and produce a backtrace that contains the missing frame(s).
```

```swift
(lldb) bt
* thread #1
  * frame #0: 0x0020392c NeoPixelRainbow`swift_UsageFault_Handler() + 28
    frame #1: 0x00203910 NeoPixelRainbow`swift_UsageFault_Handler()
    frame #2: 0x00203366 NeoPixelRainbow`static Application.main() + 2270 // Real exception location
    frame #3: 0x00202a86 NeoPixelRainbow`NeoPixelRainbow_main + 8
    frame #4: 0x00200256 NeoPixelRainbow`cinit_ctor_loop_end + 18
    frame #5: 0x00200210 NeoPixelRainbow`Reset_Handler + 16
```

## Linking

### `@section` and `@used` attributes

> [SE-0492](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0492-section-control.md) introduced two new attributes that are useful in Embedded Swift contexts. The `@section` attribute specifies that a particular global variable should be emitted into the named section:

Reviewer: ```suggestion
The Swift evolution proposal [SE-0492](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0492-section-control.md) introduced two new attributes that are useful in Embedded Swift. The `@section` attribute specifies that a particular global variable should be emitted into the named section:
```

```swift
@section("__DATA,mysection")
@used
let myLinkerSetEntry: Int = 42
```

Also included is an `objectFormat` check that can be used within `#if` to conditionalize code on the various object formats (ELF, COFF, MachO, Wasm), which can be used to provide different section names:

```swift
#if objectFormat(ELF)
@section("mysection")
#elseif objectFormat(MachO)
@section("__DATA,mysection")
#endif
var global = ...
```

> The `@used` attribute states that the entity it’s attached to should always be emitted, even if it appears unused.

Reviewer: ```suggestion
The `@used` attribute lets the compiler know that the entity it’s attached to should always be emitted, even if it appears unused.
```

### Progress on the Embedded Swift linkage model

> Embedded Swift uses a different compilation model from regular Swift that delays code generation to later in the compilation process. This compilation model has not been fully defined, and has various practical problems. One such issue involves duplicate symbols: if you have four Embedded Swift libraries whose dependencies formed a diamond, like this:

Reviewer: ```suggestion
Embedded Swift uses a different compilation model from regular Swift that delays code generation to later in the compilation process. This compilation model has not been fully defined and has various practical problems. One such issue involves duplicate symbols: if you have four Embedded Swift libraries whose dependencies formed a diamond, like this:
```
